File: lib/product/lockfile.ml
type entry = {
  provider : Ir.provider;
  reference : string;
  revision : string;
  digest : string;
  source : string;
  summary : Dependency_summary.t option;
}

type t = { schema : string; entries : entry list; integrity : string }

let compare_entry left right =
  match
    String.compare
      (Ir.provider_name left.provider)
      (Ir.provider_name right.provider)
  with
  | 0 -> String.compare left.reference right.reference
  | comparison -> comparison

let entry_json entry =
  let fields =
    [
      ("digest", Json.String entry.digest);
      ("provider", Json.String (Ir.provider_name entry.provider));
      ("reference", Json.String entry.reference);
      ("revision", Json.String entry.revision);
      ("source", Json.String entry.source);
    ]
  in
  Json.Object
    (match entry.summary with
    | None -> fields
    | Some summary -> ("summary", Dependency_summary.to_json summary) :: fields)

let unsigned_json schema entries =
  Json.Object
    [
      ("entries", Json.Array (List.map entry_json entries));
      ("schema", Json.String schema);
    ]

let assemble schema entries =
  let integrity =
    "sha256:"
    ^ Sha256.digest_string (Json.to_string (unsigned_json schema entries))
  in
  { schema; entries; integrity }

let valid_hex value =
  String.length value > 0
  && String.for_all
       (function
         | '0' .. '9' | 'a' .. 'f' -> true
         | _ -> false)
       (String.lowercase_ascii value)

let valid_sha256 value =
  Util.starts_with ~prefix:"sha256:" value
  && String.length value = 71
  && valid_hex (String.sub value 7 64)

let validate_entry entry =
  if String.trim entry.reference = "" then
    Error "lock reference must not be empty"
  else if String.trim entry.revision = "" then
    Error "lock revision must not be empty"
  else if not (valid_sha256 entry.digest) then
    Error ("invalid SHA-256 digest for " ^ entry.reference)
  else if String.trim entry.source = "" then
    Error "lock source must not be empty"
  else Ok ()

let same_payload left right =
  left.revision = right.revision
  && left.digest = right.digest && left.source = right.source
  && left.summary = right.summary

let create_with_schema schema entries =
  if schema <> "lock-v1" && schema <> "lock-v2" then
    Error ("unsupported lock schema " ^ schema)
  else if
    schema = "lock-v1"
    && List.exists (fun entry -> Option.is_some entry.summary) entries
  then Error "lock-v1 entries cannot contain semantic summaries"
  else
    let entries = List.sort compare_entry entries in
    let rec validate previous accumulator = function
      | [] -> Ok (List.rev accumulator)
      | entry :: rest -> (
          match validate_entry entry with
          | Error _ as error -> error
          | Ok () -> (
              match previous with
              | Some prior when compare_entry prior entry = 0 ->
                  if same_payload prior entry then
                    validate previous accumulator rest
                  else
                    Error
                      (Printf.sprintf "conflicting lock entries for %s:%s"
                         (Ir.provider_name entry.provider)
                         entry.reference)
              | _ -> validate (Some entry) (entry :: accumulator) rest))
    in
    match validate None [] entries with
    | Error _ as error -> error
    | Ok entries -> Ok (assemble schema entries)

let create entries = create_with_schema "lock-v2" entries
let empty = assemble "lock-v2" []

let find lock provider reference =
  List.find_opt
    (fun entry -> entry.provider = provider && entry.reference = reference)
    lock.entries

let to_json lock =
  Json.Object
    [
      ("entries", Json.Array (List.map entry_json lock.entries));
      ("integrity", Json.String lock.integrity);
      ("schema", Json.String lock.schema);
    ]

let to_canonical_json lock = Json.to_string (to_json lock) ^ "\n"

let provider_of_string = function
  | "github" -> Some Ir.Github
  | "gitlab" -> Some Gitlab
  | "azure" -> Some Azure
  | "circleci" -> Some Circleci
  | _ -> None

let required_string name json =
  match Option.bind (Json.member name json) Json.as_string with
  | Some value -> Ok value
  | None -> Error ("lock entry needs string field " ^ name)

let validate_fields ~context ~allowed json =
  match Json.as_object json with
  | None -> Error (context ^ " must be an object")
  | Some fields -> (
      let names = List.map fst fields in
      let unique = Util.deduplicate_strings names in
      if List.length unique <> List.length names then
        Error (context ^ " contains a duplicate field")
      else
        match List.find_opt (fun name -> not (List.mem name allowed)) names with
        | Some name -> Error (context ^ " contains unknown field " ^ name)
        | None -> Ok ())

let parse_entry schema json =
  let open Util in
  let* () =
    validate_fields ~context:"lock entry"
      ~allowed:
        (if schema = "lock-v1" then
           [ "provider"; "reference"; "revision"; "digest"; "source" ]
         else
           [
             "provider"; "reference"; "revision"; "digest"; "source"; "summary";
           ])
      json
  in
  let* provider_name = required_string "provider" json in
  let* provider =
    match provider_of_string provider_name with
    | Some value -> Ok value
    | None -> Error ("unknown lock provider " ^ provider_name)
  in
  let* reference = required_string "reference" json in
  let* revision = required_string "revision" json in
  let* digest = required_string "digest" json in
  let* source = required_string "source" json in
  let* summary =
    match Json.member "summary" json with
    | None -> Ok None
    | Some value ->
        let* summary = Dependency_summary.of_json value in
        Ok (Some summary)
  in
  Ok { provider; reference; revision; digest; source; summary }

let parse source =
  let open Util in
  let* json =
    match Json.parse source with
    | Ok value -> Ok value
    | Error error ->
        Error (Printf.sprintf "JSON byte %d: %s" error.offset error.message)
  in
  let* () =
    validate_fields ~context:"lockfile"
      ~allowed:[ "schema"; "integrity"; "entries" ]
      json
  in
  let* schema = required_string "schema" json in
  if schema <> "lock-v1" && schema <> "lock-v2" then
    Error ("unsupported lock schema " ^ schema)
  else
    let* integrity = required_string "integrity" json in
    let* entries_json =
      match Option.bind (Json.member "entries" json) Json.as_array with
      | Some values -> Ok values
      | None -> Error "lockfile entries must be an array"
    in
    let rec parse_entries accumulator = function
      | [] -> Ok (List.rev accumulator)
      | item :: rest ->
          let* entry = parse_entry schema item in
          parse_entries (entry :: accumulator) rest
    in
    let* entries = parse_entries [] entries_json in
    let* rebuilt = create_with_schema schema entries in
    if rebuilt.integrity <> integrity then
      Error "lockfile integrity digest mismatch"
    else Ok rebuilt

File: crates/product/src/lockfile.rs
use crate::DependencySummary;
use std::collections::{BTreeMap, BTreeSet};
use workflow_verifier_domain::Provider;
use workflow_verifier_foundation::{JsonValue, valid_content_digest};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LockEntry {
    pub provider: Provider,
    pub reference: String,
    pub revision: String,
    pub digest: String,
    pub source: String,
    pub summary: Option<DependencySummary>,
}

impl LockEntry {
    #[must_use]
    pub fn new(
        provider: Provider,
        reference: impl Into<String>,
        revision: impl Into<String>,
        digest: impl Into<String>,
        source: impl Into<String>,
    ) -> Self {
        Self {
            provider,
            reference: reference.into(),
            revision: revision.into(),
            digest: digest.into(),
            source: source.into(),
            summary: None,
        }
    }

    fn validate(&self) -> Result<(), String> {
        if self.reference.trim().is_empty() {
            Err("lock reference must not be empty".to_owned())
        } else if self.revision.trim().is_empty() {
            Err("lock revision must not be empty".to_owned())
        } else if !valid_content_digest(&self.digest) {
            Err(format!("invalid SHA-256 digest for {}", self.reference))
        } else if self.source.trim().is_empty() {
            Err("lock source must not be empty".to_owned())
        } else {
            Ok(())
        }
    }

    fn to_json(&self) -> JsonValue {
        let mut fields = BTreeMap::from([
            ("digest".to_owned(), JsonValue::String(self.digest.clone())),
            (
                "provider".to_owned(),
                JsonValue::String(self.provider.name().to_owned()),
            ),
            (
                "reference".to_owned(),
                JsonValue::String(self.reference.clone()),
            ),
            (
                "revision".to_owned(),
                JsonValue::String(self.revision.clone()),
            ),
            ("source".to_owned(), JsonValue::String(self.source.clone())),
        ]);
        if let Some(summary) = &self.summary {
            fields.insert("summary".to_owned(), summary.to_json());
        }
        JsonValue::Object(fields)
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Lockfile {
    pub schema: String,
    entries: Vec<LockEntry>,
    pub integrity: String,
}

impl Lockfile {
    /// Build a canonical lock-v2 value.
    ///
    /// # Errors
    /// Rejects invalid or conflicting provider/reference entries.
    pub fn new(entries: impl IntoIterator<Item = LockEntry>) -> Result<Self, String> {
        Self::new_with_schema("lock-v2", entries)
    }

    fn new_with_schema(
        schema: &str,
        entries: impl IntoIterator<Item = LockEntry>,
    ) -> Result<Self, String> {
        if !matches!(schema, "lock-v1" | "lock-v2") {
            return Err(format!("unsupported lock schema {schema}"));
        }
        let mut entries: Vec<_> = entries.into_iter().collect();
        entries.sort_by(|left, right| {
            (left.provider, left.reference.as_str())
                .cmp(&(right.provider, right.reference.as_str()))
        });
        for entry in &entries {
            entry.validate()?;
            if schema == "lock-v1" && entry.summary.is_some() {
                return Err("lock-v1 entries cannot contain semantic summaries".to_owned());
            }
        }
        let mut unique = Vec::with_capacity(entries.len());
        for entry in entries {
            if let Some(previous) = unique.last() {
                let previous: &LockEntry = previous;
                if previous.provider == entry.provider && previous.reference == entry.reference {
                    if previous == &entry {
                        continue;
                    }
                    return Err(format!(
                        "conflicting lock entries for {}:{}",
                        entry.provider.name(),
                        entry.reference
                    ));
                }
            }
            unique.push(entry);
        }
        let unsigned = Self::unsigned_json(schema, &unique);
        Ok(Self {
            schema: schema.to_owned(),
            entries: unique,
            integrity: unsigned.canonical_digest(),
        })
    }

    fn unsigned_json(schema: &str, entries: &[LockEntry]) -> JsonValue {
        JsonValue::Object(BTreeMap::from([
            (
                "entries".to_owned(),
                JsonValue::Array(entries.iter().map(LockEntry::to_json).collect()),
            ),
            ("schema".to_owned(), JsonValue::String(schema.to_owned())),
        ]))
    }

    #[must_use]
    pub fn entries(&self) -> &[LockEntry] {
        &self.entries
    }

    #[must_use]
    pub fn find(&self, provider: Provider, reference: &str) -> Option<&LockEntry> {
        self.entries
            .iter()
            .find(|entry| entry.provider == provider && entry.reference == reference)
    }

    #[must_use]
    pub fn verify_integrity(&self) -> bool {
        self.integrity == Self::unsigned_json(&self.schema, &self.entries).canonical_digest()
    }

    fn to_json(&self) -> JsonValue {
        JsonValue::Object(BTreeMap::from([
            (
                "entries".to_owned(),
                JsonValue::Array(self.entries.iter().map(LockEntry::to_json).collect()),
            ),
            (
                "integrity".to_owned(),
                JsonValue::String(self.integrity.clone()),
            ),
            ("schema".to_owned(), JsonValue::String(self.schema.clone())),
        ]))
    }

    #[must_use]
    pub fn to_canonical_json(&self) -> String {
        self.to_json().canonical_line()
    }

    /// Parse and authenticate strict lock-v1 or lock-v2 product JSON.
    ///
    /// # Errors
    /// Rejects malformed JSON, unknown fields, invalid entries, and integrity mismatch.
    pub fn parse(source: &str) -> Result<Self, String> {
        let root = JsonValue::parse(source).map_err(|error| error.to_string())?;
        let fields = root.exact_object("lockfile", &["entries", "integrity", "schema"])?;
        require_exact(fields, &["entries", "integrity", "schema"], "lockfile")?;
        let schema = string(fields, "schema")?;
        if !matches!(schema, "lock-v1" | "lock-v2") {
            return Err(format!("unsupported lock schema {schema}"));
        }
        let integrity = string(fields, "integrity")?;
        let values = fields
            .get("entries")
            .and_then(JsonValue::as_array)
            .ok_or_else(|| "lockfile entries must be an array".to_owned())?;
        let mut entries = Vec::with_capacity(values.len());
        for value in values {
            let allowed = if schema == "lock-v1" {
                &["digest", "provider", "reference", "revision", "source"][..]
            } else {
                &[
                    "digest",
                    "provider",
                    "reference",
                    "revision",
                    "source",
                    "summary",
                ][..]
            };
            let item = value.exact_object("lock entry", allowed)?;
            let required: BTreeSet<_> = ["digest", "provider", "reference", "revision", "source"]
                .into_iter()
                .collect();
            if !required.iter().all(|name| item.contains_key(*name)) {
                return Err("lock entry is missing a required field".to_owned());
            }
            let provider = match string(item, "provider")? {
                "github" => Provider::Github,
                "gitlab" => Provider::Gitlab,
                "azure" => Provider::Azure,
                "circleci" => Provider::Circleci,
                value => return Err(format!("unknown lock provider {value}")),
            };
            entries.push(LockEntry {
                provider,
                reference: string(item, "reference")?.to_owned(),
                revision: string(item, "revision")?.to_owned(),
                digest: string(item, "digest")?.to_owned(),
                source: string(item, "source")?.to_owned(),
                summary: item
                    .get("summary")
                    .map(DependencySummary::parse)
                    .transpose()?,
            });
        }
        let rebuilt = Self::new_with_schema(schema, entries)?;
        if rebuilt.integrity != integrity {
            return Err("lockfile integrity digest mismatch".to_owned());
        }
        Ok(rebuilt)
    }
}

fn string<'a>(fields: &'a BTreeMap<String, JsonValue>, name: &str) -> Result<&'a str, String> {
    fields
        .get(name)
        .and_then(JsonValue::as_str)
        .ok_or_else(|| format!("lockfile needs string field {name}"))
}

fn require_exact(
    fields: &BTreeMap<String, JsonValue>,
    expected: &[&str],
    context: &str,
) -> Result<(), String> {
    if fields.len() == expected.len() && expected.iter().all(|name| fields.contains_key(*name)) {
        Ok(())
    } else {
        Err(format!("{context} has missing fields"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use workflow_verifier_foundation::content_digest;

    fn valid_entry(provider: Provider, reference: &str) -> LockEntry {
        LockEntry::new(
            provider,
            reference,
            "immutable-revision",
            content_digest("locked content"),
            "https://example.test/source",
        )
    }

    #[test]
    fn entry_validation_and_lookup_require_every_identity_field() {
        let valid = valid_entry(Provider::Github, "owner/action@main");
        assert!(valid.validate().is_ok());
        for invalid in [
            LockEntry {
                reference: " ".to_owned(),
                ..valid.clone()
            },
            LockEntry {
                revision: " ".to_owned(),
                ..valid.clone()
            },
            LockEntry {
                digest: "invalid".to_owned(),
                ..valid.clone()
            },
            LockEntry {
                source: " ".to_owned(),
                ..valid.clone()
            },
        ] {
            assert!(invalid.validate().is_err());
            assert!(Lockfile::new([invalid]).is_err());
        }

        let gitlab = valid_entry(Provider::Gitlab, "owner/action@main");
        let other = valid_entry(Provider::Github, "owner/other@main");
        let lock = Lockfile::new([valid.clone(), gitlab, other]).expect("lockfile");
        assert_eq!(lock.find(Provider::Github, &valid.reference), Some(&valid));
        assert_eq!(lock.find(Provider::Gitlab, "owner/other@main"), None);
        assert_eq!(lock.find(Provider::Github, "missing"), None);
    }

    #[test]
    fn integrity_and_exact_field_contracts_reject_independent_tampering() {
        let lock =
            Lockfile::new([valid_entry(Provider::Github, "owner/action@main")]).expect("lockfile");
        assert!(lock.verify_integrity());
        let mut tampered = lock.clone();
        tampered.entries[0].revision = "other-revision".to_owned();
        assert!(!tampered.verify_integrity());

        let exact = BTreeMap::from([
            ("first".to_owned(), JsonValue::Null),
            ("second".to_owned(), JsonValue::Null),
        ]);
        assert!(require_exact(&exact, &["first", "second"], "fixture").is_ok());
        assert!(require_exact(&exact, &["first"], "fixture").is_err());
        assert!(require_exact(&exact, &["first", "other"], "fixture").is_err());
    }
}
